src/plugins/vis_type_timeseries/public/application/components/aggs/agg.tsx

```diff
@@ -51,6 +55,30 @@ export function Agg(props: AggProps) {
   const indexPattern = props.series.override_index_pattern
     ? props.series.series_index_pattern
     : props.panel.index_pattern;
+  const isKibanaIndexPattern = props.panel.use_kibana_indexes || indexPattern === '';
+
+  const onAggChange = useMemo(
+    () => seriesChangeHandler({ name, model: series, onChange }, siblings),
+    [name, onChange, siblings, series]
+  );
+
+  useEffect(() => {
+    const isNumericMetric = checkIfNumericMetric(model, fields, indexPattern);
+    const isNumberFormatter = ![DATA_FORMATTERS.DEFAULT, DATA_FORMATTERS.CUSTOM].includes(
+      series.formatter as DATA_FORMATTERS
+    );
+
+    if (isNumberFormatter && !isNumericMetric) {
+      onChange({ formatter: DATA_FORMATTERS.DEFAULT });
+    }
+    // in case of string index pattern mode, change default formatter depending on metric type
+    // "number" formatter for numeric metric and "" as custom formatter for any other type
+    if (series.formatter === DATA_FORMATTERS.DEFAULT && !isKibanaIndexPattern) {
+      onChange({
+        formatter: isNumericMetric ? DATA_FORMATTERS.NUMBER : '',
+      });
+    }
+  }, [indexPattern, model, onChange, fields, series.formatter, isKibanaIndexPattern]);
```

Comment:
This hook `series.formatter` dependency here leads to infinite react rerenders. See video with reproducing this. One of the ways to solve it might be adding `series.formatter` in one of two `if` statement condition.


https://user-images.githubusercontent.com/39378793/129922782-29d54b4c-a544-4559-9e8e-d7db819713f3.mp4